server/src/result.ts

```diff
@@ -0,0 +1,9 @@
+/// Basic Result type similar to Rust
+export type Result<T, E extends Error> = Ok<T> | E;
```

Comment:
There's a few reasons I thought a result type was important to introduce:

1. I needed a way to decouple the alerts that are sent to the LSP client from errors that occur when communicating with F*. Previously all the F* communication bits had to take the LSP client connection so they could send alert messages when errors occurred.
2. The Result type makes the kinds of errors that can be returned explicit in the type annotation (with `Error` being a catchall). This is useful because the F* handlers only send alerts on some error messages (e.g. not if an alert has already been sent as a result of a previous F* error). The explicit error types make it clear what specific kinds of errors a function may return. To my knowledge, it's not possible to annotate (outside of comments) what exceptions can be thrown, making it more implicit than if we used an explicit result type.

Given those reasons, do you think it's still better to go with exceptions instead?

Reply:
> I needed a way to decouple the alerts that are sent to the LSP client from errors that occur when communicating with F*. Previously all the F* communication bits had to take the LSP client connection so they could send alert messages when errors occurred.

IMHO the proper way to decouple this is to add an `processDied: Event<{exitCode: number}>` on `FStarConnection`.

AFAICT the only thing we can do about a dead F* process is to inform the user.  So there's no reason to have separate error handling code in every request handler.  We should just throw an exception (and thus abort).

> This is useful because the F* handlers only send alerts on some error messages [...]

Is there any other case where this happens right now except for "process died"?

> Given those reasons, do you think it's still better to go with exceptions instead?

For unrecoverable errors like "process died," I think we should definitely go with exceptions.

In other cases, I think it's better to use an ADT.  Like in the responses returned by F*, which can be either success or failure.  But there I'd just pass the typed JSON value around:
```typescript
type IdeResponse<T, E> =
  { kind: 'response', 'query-id': string, response: T, status: 'success' } |
  { kind: 'response', 'query-id': string, response: E, status: 'failure' }
```

Reply:
>AFAICT the only thing we can do about a dead F* process is to inform the user. So there's no reason to have separate error handling code in every request handler.

Yeah that makes sense to me. You could try to run a new F* process as well, but this would still be something that should be handled at a process level rather than at a request level.

> > This is useful because the F* handlers only send alerts on some error messages [...]
>
> Is there any other case where this happens right now except for "process died"?

I believe this only happens for "process died". There's also special behavior around ignoring validation queries if F* doesn't support full buffer queries, but this seems like one-off behavior that may not be relevant anymore and could just be rolled into an exception/error.